scrapy_poet/utils.py

```diff
@@ -125,40 +131,47 @@ def is_response_going_to_be_used(request, spider):
     return False
 
 
-def build_plan(callback, instances) -> Tuple[andi.Plan, Dict[Type, PageObjectInputProvider]]:
+def build_plan(callback) -> andi.Plan:
     """Build a plan for the injection in the callback."""
-    provider_instances = build_providers(instances)
-    plan = andi.plan(
+    return andi.plan(
         callback,
         is_injectable=is_injectable,
-        externally_provided=provider_instances.keys()
+        externally_provided=get_provided_classes_from_providers(),
     )
-    return plan, provider_instances
-
 
-def build_providers(instances) -> Dict[Type, PageObjectInputProvider]:
-    result = {}
-    for cls, provider in providers.items():
-        kwargs = andi.plan(
-            provider,
-            is_injectable=is_injectable,
-            externally_provided=instances.keys(),
-            full_final_kwargs=True,
-        ).final_kwargs(instances)
-        result[cls] = provider(**kwargs)  # type: ignore
 
-    return result
+def build_provider(provider: Type[PageObjectInputProvider],
+                   external_dependencies: Dict[Callable, Any]) -> PageObjectInputProvider:
+    kwargs = andi.plan(
+        provider,
+        is_injectable=is_injectable,
+        externally_provided=external_dependencies.keys(),
+        full_final_kwargs=True,
+    ).final_kwargs(external_dependencies)
+    return provider(**kwargs)  # type: ignore
 
 
 @inlineCallbacks
-def build_instances(plan: andi.Plan, providers):
-    """Build the instances dict from a plan."""
+def build_instances(plan: andi.Plan, external_dependencies: Dict[Callable, Any]):
+    """Build the instances dict from a plan including external dependencies."""
     instances = {}
-    for cls, kwargs_spec in plan:
-        if cls in providers:
-            instances[cls] = yield maybeDeferred_coro(providers[cls])
-        else:
+
+    # Build dependencies handled by registered providers
+    plan_dependencies = dict(plan.dependencies)
+    for provider in providers:
+        provided_classes = plan_dependencies.keys() & provider.provided_classes
+        if not provided_classes:
+            continue
+
+        provider_instance = build_provider(provider, external_dependencies)
+        results = yield maybeDeferred_coro(provider_instance, provided_classes)
```

Comment:
I see that you directly send in `provided_classes` these classes that the provider must deal with. Makes sense :+1: